// @eai/schema-parser/src/types/index.ts
// XML/XSD 파싱 관련 타입 정의

// 임시 타입 정의 (shared 패키지에서 가져올 예정)
export interface XSDSchema {
  targetNamespace: string;
  elementFormDefault: string;
  attributeFormDefault: string;
  elements: Element[];
  complexTypes: ComplexType[];
  simpleTypes: SimpleType[];
  attributes: Attribute[];
  groups: any[];
  attributeGroups: any[];
  imports: any[];
  includes: any[];
}

export interface ComplexType {
  name: string;
  namespace?: string;
  baseType?: string;
  elements: Element[];
  attributes: Attribute[];
  mixed?: boolean;
  abstract?: boolean;
}

export interface SimpleType {
  name: string;
  namespace?: string;
  baseType?: string;
  restrictions: any[];
  enumeration: string[];
}

export interface Element {
  name: string;
  type?: string;
  minOccurs?: number;
  maxOccurs?: number | 'unbounded';
}

export interface Attribute {
  name: string;
  type?: string;
  use?: 'required' | 'optional' | 'prohibited';
}

/**
 * XML 파싱 결과
 */
export interface XMLParseResult {
  /** 파싱된 XML 객체 */
  data: any;
  /** 파싱 에러 목록 */
  errors: ParseError[];
  /** 파싱 경고 목록 */
  warnings: ParseError[];
  /** 네임스페이스 정보 */
  namespaces: NamespaceContext;
}

/**
 * XSD 파싱 결과
 */
export interface XSDParseResult {
  /** 파싱된 XSD 스키마 */
  schema: XSDSchema;
  /** 파싱 에러 목록 */
  errors: ParseError[];
  /** 파싱 경고 목록 */
  warnings: ParseError[];
  /** 복합 타입 목록 */
  complexTypes: ComplexType[];
  /** 단순 타입 목록 */
  simpleTypes: SimpleType[];
}

/**
 * WSDL 파싱 결과 (shared 패키지에서 가져올 예정)
 */
export interface WSDLParseResult {
  /** 파싱된 WSDL 스키마 */
  schema: any;
  /** 검증 결과 */
  validation: any;
  /** 메타데이터 */
  metadata: any;
}

/**
 * 파싱 옵션
 */
export interface ParseOptions {
  /** 네임스페이스 처리 여부 */
  processNamespaces?: boolean;
  /** 참조 해결 여부 */
  resolveReferences?: boolean;
  /** 부분 실패 허용 여부 */
  allowPartialFailure?: boolean;
  /** 최대 파싱 깊이 */
  maxDepth?: number;
  /** 타임아웃 (밀리초) */
  timeout?: number;
}

/**
 * 네임스페이스 컨텍스트
 */
export interface NamespaceContext {
  /** 네임스페이스 매핑 (prefix -> URI) */
  prefixToUri: Map<string, string>;
  /** 역방향 네임스페이스 매핑 (URI -> prefix) */
  uriToPrefix: Map<string, string>;
  /** 기본 네임스페이스 */
  defaultNamespace?: string;
  /** 타겟 네임스페이스 */
  targetNamespace?: string;
}

/**
 * 파싱 에러
 */
export interface ParseError {
  /** 에러 코드 */
  code: string;
  /** 에러 메시지 */
  message: string;
  /** 에러 위치 (라인 번호) */
  line?: number;
  /** 에러 위치 (컬럼 번호) */
  column?: number;
  /** 에러가 발생한 요소 경로 */
  path?: string;
  /** 에러 심각도 */
  severity: 'error' | 'warning' | 'info';
}

/**
 * 파싱 경고
 */
export interface ParseWarning extends ParseError {
  severity: 'warning';
}

/**
 * 참조 정보
 */
export interface ReferenceInfo {
  /** 참조 타입 */
  type: 'element' | 'type' | 'group' | 'attributeGroup';
  /** 참조 이름 */
  name: string;
  /** 참조 네임스페이스 */
  namespace?: string;
  /** 참조 위치 */
  location?: string;
}

/**
 * 타입 정의 정보
 */
export interface TypeDefinitionInfo {
  /** 타입 이름 */
  name: string;
  /** 타입 종류 */
  kind: 'complexType' | 'simpleType';
  /** 네임스페이스 */
  namespace?: string;
  /** 기본 타입 */
  baseType?: string;
  /** 제약 조건 */
  restrictions?: any[];
  /** 속성 목록 */
  attributes?: any[];
  /** 요소 목록 (complexType인 경우) */
  elements?: any[];
}
