// @eai/schema-parser/dist/parsers/base-parser.js
/**
 * 모든 스키마 파서의 기본 클래스
 */
export class BaseParser {
    constructor() { }
    /**
     * 공통 유틸리티 메서드들
     */
    normalizeWhitespace(text) {
        return text.replace(/\s+/g, ' ').trim();
    }
    isValidUrl(url) {
        try {
            new URL(url);
            return true;
        }
        catch {
            return false;
        }
    }
    extractTextContent(element) {
        if (typeof element === 'string') {
            return element.trim();
        }
        if (element && element['#text']) {
            return element['#text'].trim();
        }
        return '';
    }
    ensureArray(value) {
        if (Array.isArray(value)) {
            return value;
        }
        return value ? [value] : [];
    }
    removeEmptyValues(obj) {
        if (Array.isArray(obj)) {
            return obj.map(item => this.removeEmptyValues(item)).filter(item => item !== null && item !== undefined);
        }
        if (obj && typeof obj === 'object') {
            const cleaned = {};
            Object.keys(obj).forEach(key => {
                const value = this.removeEmptyValues(obj[key]);
                if (value !== null && value !== undefined && value !== '') {
                    cleaned[key] = value;
                }
            });
            return Object.keys(cleaned).length > 0 ? cleaned : undefined;
        }
        return obj;
    }
}
