// @eai/shared/dist/types/unified.js
// 통합 스키마 모델 인터페이스
import { SchemaType } from './index';
export var ConversionImpact;
(function (ConversionImpact) {
    ConversionImpact["NONE"] = "NONE";
    ConversionImpact["LOW"] = "LOW";
    ConversionImpact["MEDIUM"] = "MEDIUM";
    ConversionImpact["HIGH"] = "HIGH";
    ConversionImpact["CRITICAL"] = "CRITICAL";
})(ConversionImpact || (ConversionImpact = {}));
export var PipelineStatus;
(function (PipelineStatus) {
    PipelineStatus["PENDING"] = "PENDING";
    PipelineStatus["RUNNING"] = "RUNNING";
    PipelineStatus["COMPLETED"] = "COMPLETED";
    PipelineStatus["FAILED"] = "FAILED";
    PipelineStatus["CANCELLED"] = "CANCELLED";
})(PipelineStatus || (PipelineStatus = {}));
export var StageStatus;
(function (StageStatus) {
    StageStatus["PENDING"] = "PENDING";
    StageStatus["RUNNING"] = "RUNNING";
    StageStatus["COMPLETED"] = "COMPLETED";
    StageStatus["FAILED"] = "FAILED";
    StageStatus["SKIPPED"] = "SKIPPED";
})(StageStatus || (StageStatus = {}));
export var DependencyType;
(function (DependencyType) {
    DependencyType["IMPORT"] = "IMPORT";
    DependencyType["INCLUDE"] = "INCLUDE";
    DependencyType["REFERENCE"] = "REFERENCE";
    DependencyType["INHERITANCE"] = "INHERITANCE";
})(DependencyType || (DependencyType = {}));
export var ReferenceType;
(function (ReferenceType) {
    ReferenceType["SCHEMA"] = "SCHEMA";
    ReferenceType["NAMESPACE"] = "NAMESPACE";
    ReferenceType["TYPE"] = "TYPE";
    ReferenceType["SERVICE"] = "SERVICE";
})(ReferenceType || (ReferenceType = {}));
export var ReferenceStatus;
(function (ReferenceStatus) {
    ReferenceStatus["VALID"] = "VALID";
    ReferenceStatus["INVALID"] = "INVALID";
    ReferenceStatus["UNREACHABLE"] = "UNREACHABLE";
    ReferenceStatus["UNKNOWN"] = "UNKNOWN";
})(ReferenceStatus || (ReferenceStatus = {}));
export var PatternCategory;
(function (PatternCategory) {
    PatternCategory["DESIGN"] = "DESIGN";
    PatternCategory["STRUCTURAL"] = "STRUCTURAL";
    PatternCategory["BEHAVIORAL"] = "BEHAVIORAL";
    PatternCategory["INTEGRATION"] = "INTEGRATION";
})(PatternCategory || (PatternCategory = {}));
export var AntiPatternSeverity;
(function (AntiPatternSeverity) {
    AntiPatternSeverity["LOW"] = "LOW";
    AntiPatternSeverity["MEDIUM"] = "MEDIUM";
    AntiPatternSeverity["HIGH"] = "HIGH";
    AntiPatternSeverity["CRITICAL"] = "CRITICAL";
})(AntiPatternSeverity || (AntiPatternSeverity = {}));
export var AnalysisRecommendationType;
(function (AnalysisRecommendationType) {
    AnalysisRecommendationType["REFACTORING"] = "REFACTORING";
    AnalysisRecommendationType["OPTIMIZATION"] = "OPTIMIZATION";
    AnalysisRecommendationType["STANDARDIZATION"] = "STANDARDIZATION";
    AnalysisRecommendationType["DOCUMENTATION"] = "DOCUMENTATION";
    AnalysisRecommendationType["SECURITY"] = "SECURITY";
    AnalysisRecommendationType["PERFORMANCE"] = "PERFORMANCE";
})(AnalysisRecommendationType || (AnalysisRecommendationType = {}));
export var AnalysisRecommendationPriority;
(function (AnalysisRecommendationPriority) {
    AnalysisRecommendationPriority["CRITICAL"] = "CRITICAL";
    AnalysisRecommendationPriority["HIGH"] = "HIGH";
    AnalysisRecommendationPriority["MEDIUM"] = "MEDIUM";
    AnalysisRecommendationPriority["LOW"] = "LOW";
})(AnalysisRecommendationPriority || (AnalysisRecommendationPriority = {}));
export var EffortComplexity;
(function (EffortComplexity) {
    EffortComplexity["TRIVIAL"] = "TRIVIAL";
    EffortComplexity["SIMPLE"] = "SIMPLE";
    EffortComplexity["MODERATE"] = "MODERATE";
    EffortComplexity["COMPLEX"] = "COMPLEX";
    EffortComplexity["VERY_COMPLEX"] = "VERY_COMPLEX";
})(EffortComplexity || (EffortComplexity = {}));
export var ImpactLevel;
(function (ImpactLevel) {
    ImpactLevel["NONE"] = "NONE";
    ImpactLevel["LOW"] = "LOW";
    ImpactLevel["MEDIUM"] = "MEDIUM";
    ImpactLevel["HIGH"] = "HIGH";
    ImpactLevel["VERY_HIGH"] = "VERY_HIGH";
})(ImpactLevel || (ImpactLevel = {}));
export var StandardCategory;
(function (StandardCategory) {
    StandardCategory["INDUSTRY"] = "INDUSTRY";
    StandardCategory["ORGANIZATIONAL"] = "ORGANIZATIONAL";
    StandardCategory["TECHNICAL"] = "TECHNICAL";
    StandardCategory["SECURITY"] = "SECURITY";
    StandardCategory["REGULATORY"] = "REGULATORY";
})(StandardCategory || (StandardCategory = {}));
export var ViolationSeverity;
(function (ViolationSeverity) {
    ViolationSeverity["MINOR"] = "MINOR";
    ViolationSeverity["MAJOR"] = "MAJOR";
    ViolationSeverity["CRITICAL"] = "CRITICAL";
    ViolationSeverity["BLOCKER"] = "BLOCKER";
})(ViolationSeverity || (ViolationSeverity = {}));
