// @eai/shared/dist/types/sap.js
export var Multiplicity;
(function (Multiplicity) {
    Multiplicity["ZERO_OR_ONE"] = "0..1";
    Multiplicity["ONE"] = "1";
    Multiplicity["MANY"] = "*";
})(Multiplicity || (Multiplicity = {}));
export var ParameterMode;
(function (ParameterMode) {
    ParameterMode["IN"] = "In";
    ParameterMode["OUT"] = "Out";
    ParameterMode["INOUT"] = "InOut";
})(ParameterMode || (ParameterMode = {}));
export var SegmentStatus;
(function (SegmentStatus) {
    SegmentStatus["MANDATORY"] = "M";
    SegmentStatus["CONDITIONAL"] = "C";
    SegmentStatus["OPTIONAL"] = "O";
})(SegmentStatus || (SegmentStatus = {}));
export var RFCParameterType;
(function (RFCParameterType) {
    RFCParameterType["ELEMENTARY"] = "ELEMENTARY";
    RFCParameterType["STRUCTURE"] = "STRUCTURE";
    RFCParameterType["TABLE"] = "TABLE";
})(RFCParameterType || (RFCParameterType = {}));
export var SAPObjectType;
(function (SAPObjectType) {
    SAPObjectType["ODATA_SERVICE"] = "ODATA_SERVICE";
    SAPObjectType["IDOC_TYPE"] = "IDOC_TYPE";
    SAPObjectType["RFC_FUNCTION"] = "RFC_FUNCTION";
    SAPObjectType["TABLE"] = "TABLE";
    SAPObjectType["VIEW"] = "VIEW";
    SAPObjectType["STRUCTURE"] = "STRUCTURE";
})(SAPObjectType || (SAPObjectType = {}));
export var SAPErrorType;
(function (SAPErrorType) {
    SAPErrorType["CONNECTION_ERROR"] = "CONNECTION_ERROR";
    SAPErrorType["AUTHENTICATION_ERROR"] = "AUTHENTICATION_ERROR";
    SAPErrorType["AUTHORIZATION_ERROR"] = "AUTHORIZATION_ERROR";
    SAPErrorType["OBJECT_NOT_FOUND"] = "OBJECT_NOT_FOUND";
    SAPErrorType["METADATA_ERROR"] = "METADATA_ERROR";
    SAPErrorType["PARSING_ERROR"] = "PARSING_ERROR";
    SAPErrorType["TIMEOUT_ERROR"] = "TIMEOUT_ERROR";
})(SAPErrorType || (SAPErrorType = {}));
