// @eai/shared/dist/types/xsd.js
export var XSDFormChoice;
(function (XSDFormChoice) {
    XSDFormChoice["QUALIFIED"] = "qualified";
    XSDFormChoice["UNQUALIFIED"] = "unqualified";
})(XSDFormChoice || (XSDFormChoice = {}));
export var XSDUse;
(function (XSDUse) {
    XSDUse["REQUIRED"] = "required";
    XSDUse["OPTIONAL"] = "optional";
    XSDUse["PROHIBITED"] = "prohibited";
})(XSDUse || (XSDUse = {}));
export var XSDContentType;
(function (XSDContentType) {
    XSDContentType["EXTENSION"] = "extension";
    XSDContentType["RESTRICTION"] = "restriction";
    XSDContentType["SIMPLE_CONTENT"] = "simpleContent";
    XSDContentType["COMPLEX_CONTENT"] = "complexContent";
})(XSDContentType || (XSDContentType = {}));
export var XSDWhiteSpace;
(function (XSDWhiteSpace) {
    XSDWhiteSpace["PRESERVE"] = "preserve";
    XSDWhiteSpace["REPLACE"] = "replace";
    XSDWhiteSpace["COLLAPSE"] = "collapse";
})(XSDWhiteSpace || (XSDWhiteSpace = {}));
export var XSDFacetType;
(function (XSDFacetType) {
    XSDFacetType["ENUMERATION"] = "enumeration";
    XSDFacetType["FRACTION_DIGITS"] = "fractionDigits";
    XSDFacetType["LENGTH"] = "length";
    XSDFacetType["MAX_EXCLUSIVE"] = "maxExclusive";
    XSDFacetType["MAX_INCLUSIVE"] = "maxInclusive";
    XSDFacetType["MAX_LENGTH"] = "maxLength";
    XSDFacetType["MIN_EXCLUSIVE"] = "minExclusive";
    XSDFacetType["MIN_INCLUSIVE"] = "minInclusive";
    XSDFacetType["MIN_LENGTH"] = "minLength";
    XSDFacetType["PATTERN"] = "pattern";
    XSDFacetType["TOTAL_DIGITS"] = "totalDigits";
    XSDFacetType["WHITE_SPACE"] = "whiteSpace";
})(XSDFacetType || (XSDFacetType = {}));
export var XSDProcessContents;
(function (XSDProcessContents) {
    XSDProcessContents["STRICT"] = "strict";
    XSDProcessContents["LAX"] = "lax";
    XSDProcessContents["SKIP"] = "skip";
})(XSDProcessContents || (XSDProcessContents = {}));
export var XSDErrorCode;
(function (XSDErrorCode) {
    XSDErrorCode["INVALID_TYPE_REFERENCE"] = "INVALID_TYPE_REFERENCE";
    XSDErrorCode["CIRCULAR_TYPE_DEFINITION"] = "CIRCULAR_TYPE_DEFINITION";
    XSDErrorCode["INVALID_FACET_VALUE"] = "INVALID_FACET_VALUE";
    XSDErrorCode["CONFLICTING_CONSTRAINTS"] = "CONFLICTING_CONSTRAINTS";
    XSDErrorCode["INVALID_NAMESPACE"] = "INVALID_NAMESPACE";
    XSDErrorCode["MISSING_REQUIRED_ATTRIBUTE"] = "MISSING_REQUIRED_ATTRIBUTE";
    XSDErrorCode["INVALID_OCCURRENCE_CONSTRAINT"] = "INVALID_OCCURRENCE_CONSTRAINT";
    XSDErrorCode["DUPLICATE_DEFINITION"] = "DUPLICATE_DEFINITION";
})(XSDErrorCode || (XSDErrorCode = {}));
