// @eai/shared/dist/types/openapi.js
export var OpenAPIParameterLocation;
(function (OpenAPIParameterLocation) {
    OpenAPIParameterLocation["QUERY"] = "query";
    OpenAPIParameterLocation["HEADER"] = "header";
    OpenAPIParameterLocation["PATH"] = "path";
    OpenAPIParameterLocation["COOKIE"] = "cookie";
})(OpenAPIParameterLocation || (OpenAPIParameterLocation = {}));
export var OpenAPIParameterStyle;
(function (OpenAPIParameterStyle) {
    OpenAPIParameterStyle["MATRIX"] = "matrix";
    OpenAPIParameterStyle["LABEL"] = "label";
    OpenAPIParameterStyle["FORM"] = "form";
    OpenAPIParameterStyle["SIMPLE"] = "simple";
    OpenAPIParameterStyle["SPACE_DELIMITED"] = "spaceDelimited";
    OpenAPIParameterStyle["PIPE_DELIMITED"] = "pipeDelimited";
    OpenAPIParameterStyle["DEEP_OBJECT"] = "deepObject";
})(OpenAPIParameterStyle || (OpenAPIParameterStyle = {}));
export var OpenAPIDataType;
(function (OpenAPIDataType) {
    OpenAPIDataType["NULL"] = "null";
    OpenAPIDataType["BOOLEAN"] = "boolean";
    OpenAPIDataType["OBJECT"] = "object";
    OpenAPIDataType["ARRAY"] = "array";
    OpenAPIDataType["NUMBER"] = "number";
    OpenAPIDataType["STRING"] = "string";
    OpenAPIDataType["INTEGER"] = "integer";
})(OpenAPIDataType || (OpenAPIDataType = {}));
export var OpenAPISecuritySchemeType;
(function (OpenAPISecuritySchemeType) {
    OpenAPISecuritySchemeType["API_KEY"] = "apiKey";
    OpenAPISecuritySchemeType["HTTP"] = "http";
    OpenAPISecuritySchemeType["OAUTH2"] = "oauth2";
    OpenAPISecuritySchemeType["OPEN_ID_CONNECT"] = "openIdConnect";
})(OpenAPISecuritySchemeType || (OpenAPISecuritySchemeType = {}));
export var OpenAPISecuritySchemeIn;
(function (OpenAPISecuritySchemeIn) {
    OpenAPISecuritySchemeIn["QUERY"] = "query";
    OpenAPISecuritySchemeIn["HEADER"] = "header";
    OpenAPISecuritySchemeIn["COOKIE"] = "cookie";
})(OpenAPISecuritySchemeIn || (OpenAPISecuritySchemeIn = {}));
export var OpenAPIErrorCode;
(function (OpenAPIErrorCode) {
    OpenAPIErrorCode["INVALID_REFERENCE"] = "INVALID_REFERENCE";
    OpenAPIErrorCode["MISSING_REQUIRED_PROPERTY"] = "MISSING_REQUIRED_PROPERTY";
    OpenAPIErrorCode["INVALID_SCHEMA"] = "INVALID_SCHEMA";
    OpenAPIErrorCode["INVALID_EXAMPLE"] = "INVALID_EXAMPLE";
    OpenAPIErrorCode["SECURITY_SCHEME_NOT_FOUND"] = "SECURITY_SCHEME_NOT_FOUND";
    OpenAPIErrorCode["INVALID_OPERATION_ID"] = "INVALID_OPERATION_ID";
    OpenAPIErrorCode["DUPLICATE_OPERATION_ID"] = "DUPLICATE_OPERATION_ID";
    OpenAPIErrorCode["INVALID_PATH_PARAMETER"] = "INVALID_PATH_PARAMETER";
})(OpenAPIErrorCode || (OpenAPIErrorCode = {}));
