// @eai/shared/src/constants/index.ts
// 애플리케이션 상수 정의

// 파일 업로드 관련 상수
export const FILE_UPLOAD = {
  MAX_SIZE: 50 * 1024 * 1024, // 50MB
  ALLOWED_MIME_TYPES: [
    'text/xml',
    'application/xml',
    'application/json',
    'text/plain',
    'application/x-yaml',
    'text/yaml',
    'application/wsdl+xml',
    'application/xsd+xml',
  ],
  ALLOWED_EXTENSIONS: ['.xml', '.wsdl', '.xsd', '.json', '.yaml', '.yml'],
  TIMEOUT: 30000, // 30초
} as const;

// URL 요청 관련 상수
export const URL_REQUEST = {
  TIMEOUT: 5000, // 5초
  MAX_REDIRECTS: 5,
  USER_AGENT: 'EAI-Schema-Toolkit/1.0',
  BLOCKED_HOSTS: ['localhost', '127.0.0.1', '0.0.0.0', '::1'],
  ALLOWED_PROTOCOLS: ['http:', 'https:'],
} as const;

// 파싱 관련 상수
export const PARSING = {
  MAX_DEPTH: 100,
  MAX_ELEMENTS: 10000,
  MAX_ATTRIBUTES: 1000,
  NAMESPACE_SEPARATOR: ':',
  DEFAULT_NAMESPACE: 'default',
} as const;

// 검증 관련 상수
export const VALIDATION = {
  MAX_ERRORS: 100,
  MAX_WARNINGS: 200,
  STRICT_MODE: false,
  ENABLE_WARNINGS: true,
} as const;

// 변환 관련 상수
export const CONVERSION = {
  DEFAULT_VERSION: '1.0',
  MAX_CONVERSION_TIME: 60000, // 60초
  BATCH_SIZE: 1000,
} as const;

// 매핑 관련 상수
export const MAPPING = {
  AUTO_MAPPING_THRESHOLD: 0.8, // 80% 유사도
  MAX_MAPPINGS: 10000,
  DEFAULT_TRANSFORM_TIMEOUT: 5000,
} as const;

// 비교 관련 상수
export const COMPARISON = {
  MAX_DIFFERENCES: 1000,
  SIMILARITY_THRESHOLD: 0.7,
  DEEP_COMPARISON_LIMIT: 1000,
} as const;

// 세션 관리 상수
export const SESSION = {
  MAX_SESSIONS: 100,
  SESSION_TIMEOUT: 24 * 60 * 60 * 1000, // 24시간
  CLEANUP_INTERVAL: 60 * 60 * 1000, // 1시간
  MAX_HISTORY_SIZE: 1000,
} as const;

// 캐시 관련 상수
export const CACHE = {
  DEFAULT_TTL: 60 * 60 * 1000, // 1시간
  MAX_CACHE_SIZE: 100 * 1024 * 1024, // 100MB
  CLEANUP_THRESHOLD: 0.8, // 80% 사용률에서 정리 시작
} as const;

// API 관련 상수
export const API = {
  VERSION: 'v1',
  BASE_PATH: '/api',
  RATE_LIMIT: {
    WINDOW_MS: 15 * 60 * 1000, // 15분
    MAX_REQUESTS: 100,
  },
  CORS: {
    ORIGIN: ['http://localhost:3000', 'http://localhost:3001'],
    METHODS: ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'],
    ALLOWED_HEADERS: ['Content-Type', 'Authorization', 'X-Requested-With'],
  },
} as const;

// 에러 코드
export const ERROR_CODES = {
  // 파일 업로드 에러
  FILE_TOO_LARGE: 'FILE_TOO_LARGE',
  INVALID_FILE_TYPE: 'INVALID_FILE_TYPE',
  UPLOAD_FAILED: 'UPLOAD_FAILED',
  
  // 파싱 에러
  PARSING_FAILED: 'PARSING_FAILED',
  INVALID_XML: 'INVALID_XML',
  INVALID_JSON: 'INVALID_JSON',
  NAMESPACE_ERROR: 'NAMESPACE_ERROR',
  REFERENCE_ERROR: 'REFERENCE_ERROR',
  
  // 검증 에러
  VALIDATION_FAILED: 'VALIDATION_FAILED',
  SCHEMA_INVALID: 'SCHEMA_INVALID',
  TYPE_MISMATCH: 'TYPE_MISMATCH',
  CONSTRAINT_VIOLATION: 'CONSTRAINT_VIOLATION',
  
  // 변환 에러
  CONVERSION_FAILED: 'CONVERSION_FAILED',
  UNSUPPORTED_CONVERSION: 'UNSUPPORTED_CONVERSION',
  CONVERSION_TIMEOUT: 'CONVERSION_TIMEOUT',
  
  // 매핑 에러
  MAPPING_FAILED: 'MAPPING_FAILED',
  INVALID_MAPPING: 'INVALID_MAPPING',
  MAPPING_CONFLICT: 'MAPPING_CONFLICT',
  
  // 비교 에러
  COMPARISON_FAILED: 'COMPARISON_FAILED',
  SCHEMA_NOT_FOUND: 'SCHEMA_NOT_FOUND',
  
  // SAP 관련 에러
  SAP_CONNECTION_FAILED: 'SAP_CONNECTION_FAILED',
  SAP_AUTH_FAILED: 'SAP_AUTH_FAILED',
  SAP_OBJECT_NOT_FOUND: 'SAP_OBJECT_NOT_FOUND',
  SAP_METADATA_ERROR: 'SAP_METADATA_ERROR',
  
  // 시스템 에러
  INTERNAL_ERROR: 'INTERNAL_ERROR',
  TIMEOUT: 'TIMEOUT',
  MEMORY_LIMIT_EXCEEDED: 'MEMORY_LIMIT_EXCEEDED',
  RATE_LIMIT_EXCEEDED: 'RATE_LIMIT_EXCEEDED',
} as const;

// 로그 레벨
export const LOG_LEVELS = {
  ERROR: 0,
  WARN: 1,
  INFO: 2,
  DEBUG: 3,
} as const;

// 정규 표현식
export const REGEX = {
  // XML 네임스페이스
  NAMESPACE: /^[a-zA-Z_][\w.-]*$/,
  
  // URL 검증
  URL: /^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)$/,
  
  // 파일명 검증
  FILENAME: /^[a-zA-Z0-9._-]+$/,
  
  // 버전 번호
  VERSION: /^\d+\.\d+(\.\d+)?$/,
  
  // 식별자
  IDENTIFIER: /^[a-zA-Z_][a-zA-Z0-9_]*$/,
  
  // XPath
  XPATH: /^(\/[a-zA-Z_][\w.-]*(\[[^\]]*\])?)+$/,
} as const;

// MIME 타입 매핑
export const MIME_TYPE_MAPPING = {
  '.xml': 'text/xml',
  '.wsdl': 'application/wsdl+xml',
  '.xsd': 'application/xsd+xml',
  '.json': 'application/json',
  '.yaml': 'application/x-yaml',
  '.yml': 'application/x-yaml',
} as const;

// 스키마 타입별 기본 설정
export const SCHEMA_DEFAULTS = {
  WSDL: {
    version: '1.1',
    namespace: 'http://schemas.xmlsoap.org/wsdl/',
    soapNamespace: 'http://schemas.xmlsoap.org/wsdl/soap/',
  },
  XSD: {
    version: '1.0',
    namespace: 'http://www.w3.org/2001/XMLSchema',
  },
  JSON_SCHEMA: {
    version: 'draft-07',
    schema: 'http://json-schema.org/draft-07/schema#',
  },
  OPENAPI: {
    version: '3.0.0',
    infoVersion: '1.0.0',
  },
  SAP_ODATA: {
    version: '1.0',
    namespace: 'http://schemas.microsoft.com/ado/2008/09/edm',
  },
} as const;

// 성능 관련 상수
export const PERFORMANCE = {
  CHUNK_SIZE: 1024 * 1024, // 1MB
  MAX_MEMORY_USAGE: 512 * 1024 * 1024, // 512MB
  GC_THRESHOLD: 0.8, // 80% 메모리 사용률에서 GC 실행
  WORKER_POOL_SIZE: 4,
} as const;
