// @eai/shared/dist/types/errors.js
// 에러 심각도
export var ErrorSeverity;
(function (ErrorSeverity) {
    ErrorSeverity["CRITICAL"] = "CRITICAL";
    ErrorSeverity["ERROR"] = "ERROR";
    ErrorSeverity["WARNING"] = "WARNING";
    ErrorSeverity["INFO"] = "INFO";
    ErrorSeverity["DEBUG"] = "DEBUG";
})(ErrorSeverity || (ErrorSeverity = {}));
// 에러 카테고리
export var ErrorCategory;
(function (ErrorCategory) {
    ErrorCategory["PARSING"] = "PARSING";
    ErrorCategory["VALIDATION"] = "VALIDATION";
    ErrorCategory["CONVERSION"] = "CONVERSION";
    ErrorCategory["MAPPING"] = "MAPPING";
    ErrorCategory["COMPARISON"] = "COMPARISON";
    ErrorCategory["NETWORK"] = "NETWORK";
    ErrorCategory["FILE_SYSTEM"] = "FILE_SYSTEM";
    ErrorCategory["SECURITY"] = "SECURITY";
    ErrorCategory["CONFIGURATION"] = "CONFIGURATION";
    ErrorCategory["SYSTEM"] = "SYSTEM";
})(ErrorCategory || (ErrorCategory = {}));
export var ParsePhase;
(function (ParsePhase) {
    ParsePhase["LEXICAL_ANALYSIS"] = "LEXICAL_ANALYSIS";
    ParsePhase["SYNTAX_ANALYSIS"] = "SYNTAX_ANALYSIS";
    ParsePhase["SEMANTIC_ANALYSIS"] = "SEMANTIC_ANALYSIS";
    ParsePhase["REFERENCE_RESOLUTION"] = "REFERENCE_RESOLUTION";
    ParsePhase["TYPE_CHECKING"] = "TYPE_CHECKING";
    ParsePhase["VALIDATION"] = "VALIDATION";
})(ParsePhase || (ParsePhase = {}));
export var ValidationType;
(function (ValidationType) {
    ValidationType["SYNTAX"] = "SYNTAX";
    ValidationType["SEMANTIC"] = "SEMANTIC";
    ValidationType["REFERENCE_INTEGRITY"] = "REFERENCE_INTEGRITY";
    ValidationType["TYPE_COMPATIBILITY"] = "TYPE_COMPATIBILITY";
    ValidationType["CONSTRAINT_VIOLATION"] = "CONSTRAINT_VIOLATION";
    ValidationType["NAMESPACE_VALIDATION"] = "NAMESPACE_VALIDATION";
    ValidationType["SCHEMA_COMPLIANCE"] = "SCHEMA_COMPLIANCE";
    ValidationType["BUSINESS_RULE"] = "BUSINESS_RULE";
})(ValidationType || (ValidationType = {}));
export var ConversionPhase;
(function (ConversionPhase) {
    ConversionPhase["PRE_PROCESSING"] = "PRE_PROCESSING";
    ConversionPhase["TYPE_MAPPING"] = "TYPE_MAPPING";
    ConversionPhase["STRUCTURE_TRANSFORMATION"] = "STRUCTURE_TRANSFORMATION";
    ConversionPhase["CONSTRAINT_MAPPING"] = "CONSTRAINT_MAPPING";
    ConversionPhase["POST_PROCESSING"] = "POST_PROCESSING";
    ConversionPhase["VALIDATION"] = "VALIDATION";
})(ConversionPhase || (ConversionPhase = {}));
export var MappingType;
(function (MappingType) {
    MappingType["FIELD_MAPPING"] = "FIELD_MAPPING";
    MappingType["TYPE_MAPPING"] = "TYPE_MAPPING";
    MappingType["CONSTRAINT_MAPPING"] = "CONSTRAINT_MAPPING";
    MappingType["TRANSFORMATION_MAPPING"] = "TRANSFORMATION_MAPPING";
})(MappingType || (MappingType = {}));
export var ComparisonType;
(function (ComparisonType) {
    ComparisonType["STRUCTURE_COMPARISON"] = "STRUCTURE_COMPARISON";
    ComparisonType["TYPE_COMPARISON"] = "TYPE_COMPARISON";
    ComparisonType["CONSTRAINT_COMPARISON"] = "CONSTRAINT_COMPARISON";
    ComparisonType["SEMANTIC_COMPARISON"] = "SEMANTIC_COMPARISON";
})(ComparisonType || (ComparisonType = {}));
export var FileOperation;
(function (FileOperation) {
    FileOperation["READ"] = "READ";
    FileOperation["WRITE"] = "WRITE";
    FileOperation["DELETE"] = "DELETE";
    FileOperation["CREATE"] = "CREATE";
    FileOperation["MOVE"] = "MOVE";
    FileOperation["COPY"] = "COPY";
})(FileOperation || (FileOperation = {}));
export var SecurityType;
(function (SecurityType) {
    SecurityType["XXE_ATTACK"] = "XXE_ATTACK";
    SecurityType["SSRF_ATTACK"] = "SSRF_ATTACK";
    SecurityType["FILE_UPLOAD_VIOLATION"] = "FILE_UPLOAD_VIOLATION";
    SecurityType["SIZE_LIMIT_EXCEEDED"] = "SIZE_LIMIT_EXCEEDED";
    SecurityType["MALICIOUS_CONTENT"] = "MALICIOUS_CONTENT";
    SecurityType["UNAUTHORIZED_ACCESS"] = "UNAUTHORIZED_ACCESS";
})(SecurityType || (SecurityType = {}));
export var ThreatLevel;
(function (ThreatLevel) {
    ThreatLevel["LOW"] = "LOW";
    ThreatLevel["MEDIUM"] = "MEDIUM";
    ThreatLevel["HIGH"] = "HIGH";
    ThreatLevel["CRITICAL"] = "CRITICAL";
})(ThreatLevel || (ThreatLevel = {}));
export var ResourceType;
(function (ResourceType) {
    ResourceType["MEMORY"] = "MEMORY";
    ResourceType["CPU"] = "CPU";
    ResourceType["DISK"] = "DISK";
    ResourceType["NETWORK"] = "NETWORK";
    ResourceType["DATABASE"] = "DATABASE";
})(ResourceType || (ResourceType = {}));
export var ValidationRuleResult;
(function (ValidationRuleResult) {
    ValidationRuleResult["PASSED"] = "PASSED";
    ValidationRuleResult["FAILED"] = "FAILED";
    ValidationRuleResult["SKIPPED"] = "SKIPPED";
    ValidationRuleResult["ERROR"] = "ERROR";
})(ValidationRuleResult || (ValidationRuleResult = {}));
export var RecommendationType;
(function (RecommendationType) {
    RecommendationType["FIX"] = "FIX";
    RecommendationType["IMPROVEMENT"] = "IMPROVEMENT";
    RecommendationType["BEST_PRACTICE"] = "BEST_PRACTICE";
    RecommendationType["PERFORMANCE"] = "PERFORMANCE";
    RecommendationType["SECURITY"] = "SECURITY";
})(RecommendationType || (RecommendationType = {}));
export var RecommendationPriority;
(function (RecommendationPriority) {
    RecommendationPriority["CRITICAL"] = "CRITICAL";
    RecommendationPriority["HIGH"] = "HIGH";
    RecommendationPriority["MEDIUM"] = "MEDIUM";
    RecommendationPriority["LOW"] = "LOW";
})(RecommendationPriority || (RecommendationPriority = {}));
export var ErrorAction;
(function (ErrorAction) {
    ErrorAction["FAIL_FAST"] = "FAIL_FAST";
    ErrorAction["CONTINUE"] = "CONTINUE";
    ErrorAction["RETRY"] = "RETRY";
    ErrorAction["FALLBACK"] = "FALLBACK";
    ErrorAction["IGNORE"] = "IGNORE";
    ErrorAction["LOG_ONLY"] = "LOG_ONLY";
})(ErrorAction || (ErrorAction = {}));
export var FallbackType;
(function (FallbackType) {
    FallbackType["DEFAULT_VALUE"] = "DEFAULT_VALUE";
    FallbackType["SKIP_ELEMENT"] = "SKIP_ELEMENT";
    FallbackType["PARTIAL_PROCESSING"] = "PARTIAL_PROCESSING";
    FallbackType["ALTERNATIVE_PARSER"] = "ALTERNATIVE_PARSER";
    FallbackType["MANUAL_INTERVENTION"] = "MANUAL_INTERVENTION";
})(FallbackType || (FallbackType = {}));
