// @eai/shared/dist/types/json-schema.js
export var JSONSchemaType;
(function (JSONSchemaType) {
    JSONSchemaType["NULL"] = "null";
    JSONSchemaType["BOOLEAN"] = "boolean";
    JSONSchemaType["OBJECT"] = "object";
    JSONSchemaType["ARRAY"] = "array";
    JSONSchemaType["NUMBER"] = "number";
    JSONSchemaType["STRING"] = "string";
    JSONSchemaType["INTEGER"] = "integer";
})(JSONSchemaType || (JSONSchemaType = {}));
export var JSONSchemaFormat;
(function (JSONSchemaFormat) {
    // String formats
    JSONSchemaFormat["DATE_TIME"] = "date-time";
    JSONSchemaFormat["DATE"] = "date";
    JSONSchemaFormat["TIME"] = "time";
    JSONSchemaFormat["DURATION"] = "duration";
    JSONSchemaFormat["EMAIL"] = "email";
    JSONSchemaFormat["IDN_EMAIL"] = "idn-email";
    JSONSchemaFormat["HOSTNAME"] = "hostname";
    JSONSchemaFormat["IDN_HOSTNAME"] = "idn-hostname";
    JSONSchemaFormat["IPV4"] = "ipv4";
    JSONSchemaFormat["IPV6"] = "ipv6";
    JSONSchemaFormat["URI"] = "uri";
    JSONSchemaFormat["URI_REFERENCE"] = "uri-reference";
    JSONSchemaFormat["IRI"] = "iri";
    JSONSchemaFormat["IRI_REFERENCE"] = "iri-reference";
    JSONSchemaFormat["UUID"] = "uuid";
    JSONSchemaFormat["URI_TEMPLATE"] = "uri-template";
    JSONSchemaFormat["JSON_POINTER"] = "json-pointer";
    JSONSchemaFormat["RELATIVE_JSON_POINTER"] = "relative-json-pointer";
    JSONSchemaFormat["REGEX"] = "regex";
    // Numeric formats
    JSONSchemaFormat["INT32"] = "int32";
    JSONSchemaFormat["INT64"] = "int64";
    JSONSchemaFormat["FLOAT"] = "float";
    JSONSchemaFormat["DOUBLE"] = "double";
    // Binary formats
    JSONSchemaFormat["BYTE"] = "byte";
    JSONSchemaFormat["BINARY"] = "binary";
    JSONSchemaFormat["BASE64"] = "base64";
    JSONSchemaFormat["BASE64URL"] = "base64url";
    // Password
    JSONSchemaFormat["PASSWORD"] = "password";
})(JSONSchemaFormat || (JSONSchemaFormat = {}));
export var JSONSchemaDraft;
(function (JSONSchemaDraft) {
    JSONSchemaDraft["DRAFT_04"] = "http://json-schema.org/draft-04/schema#";
    JSONSchemaDraft["DRAFT_06"] = "http://json-schema.org/draft-06/schema#";
    JSONSchemaDraft["DRAFT_07"] = "http://json-schema.org/draft-07/schema#";
    JSONSchemaDraft["DRAFT_2019_09"] = "https://json-schema.org/draft/2019-09/schema";
    JSONSchemaDraft["DRAFT_2020_12"] = "https://json-schema.org/draft/2020-12/schema";
})(JSONSchemaDraft || (JSONSchemaDraft = {}));
export var JSONSchemaErrorCode;
(function (JSONSchemaErrorCode) {
    JSONSchemaErrorCode["INVALID_REFERENCE"] = "INVALID_REFERENCE";
    JSONSchemaErrorCode["CIRCULAR_REFERENCE"] = "CIRCULAR_REFERENCE";
    JSONSchemaErrorCode["INVALID_TYPE"] = "INVALID_TYPE";
    JSONSchemaErrorCode["INVALID_FORMAT"] = "INVALID_FORMAT";
    JSONSchemaErrorCode["CONSTRAINT_VIOLATION"] = "CONSTRAINT_VIOLATION";
    JSONSchemaErrorCode["MISSING_REQUIRED_PROPERTY"] = "MISSING_REQUIRED_PROPERTY";
    JSONSchemaErrorCode["ADDITIONAL_PROPERTIES_NOT_ALLOWED"] = "ADDITIONAL_PROPERTIES_NOT_ALLOWED";
    JSONSchemaErrorCode["ENUM_MISMATCH"] = "ENUM_MISMATCH";
    JSONSchemaErrorCode["CONST_MISMATCH"] = "CONST_MISMATCH";
    JSONSchemaErrorCode["PATTERN_MISMATCH"] = "PATTERN_MISMATCH";
    JSONSchemaErrorCode["MULTIPLE_OF_VIOLATION"] = "MULTIPLE_OF_VIOLATION";
    JSONSchemaErrorCode["MINIMUM_VIOLATION"] = "MINIMUM_VIOLATION";
    JSONSchemaErrorCode["MAXIMUM_VIOLATION"] = "MAXIMUM_VIOLATION";
    JSONSchemaErrorCode["MIN_LENGTH_VIOLATION"] = "MIN_LENGTH_VIOLATION";
    JSONSchemaErrorCode["MAX_LENGTH_VIOLATION"] = "MAX_LENGTH_VIOLATION";
    JSONSchemaErrorCode["MIN_ITEMS_VIOLATION"] = "MIN_ITEMS_VIOLATION";
    JSONSchemaErrorCode["MAX_ITEMS_VIOLATION"] = "MAX_ITEMS_VIOLATION";
    JSONSchemaErrorCode["UNIQUE_ITEMS_VIOLATION"] = "UNIQUE_ITEMS_VIOLATION";
    JSONSchemaErrorCode["MIN_PROPERTIES_VIOLATION"] = "MIN_PROPERTIES_VIOLATION";
    JSONSchemaErrorCode["MAX_PROPERTIES_VIOLATION"] = "MAX_PROPERTIES_VIOLATION";
    JSONSchemaErrorCode["DEPENDENCY_VIOLATION"] = "DEPENDENCY_VIOLATION";
    JSONSchemaErrorCode["CONDITIONAL_SCHEMA_VIOLATION"] = "CONDITIONAL_SCHEMA_VIOLATION";
    JSONSchemaErrorCode["COMPOSITION_VIOLATION"] = "COMPOSITION_VIOLATION";
})(JSONSchemaErrorCode || (JSONSchemaErrorCode = {}));
export var DocumentationFormat;
(function (DocumentationFormat) {
    DocumentationFormat["HTML"] = "html";
    DocumentationFormat["MARKDOWN"] = "markdown";
    DocumentationFormat["JSON"] = "json";
    DocumentationFormat["YAML"] = "yaml";
})(DocumentationFormat || (DocumentationFormat = {}));
