// @eai/shared/dist/types/wsdl.js
export var WSDLAddressType;
(function (WSDLAddressType) {
    WSDLAddressType["SOAP"] = "soap:address";
    WSDLAddressType["HTTP"] = "http:address";
    WSDLAddressType["MIME"] = "mime:address";
})(WSDLAddressType || (WSDLAddressType = {}));
export var WSDLStyle;
(function (WSDLStyle) {
    WSDLStyle["DOCUMENT"] = "document";
    WSDLStyle["RPC"] = "rpc";
})(WSDLStyle || (WSDLStyle = {}));
export var WSDLUse;
(function (WSDLUse) {
    WSDLUse["LITERAL"] = "literal";
    WSDLUse["ENCODED"] = "encoded";
})(WSDLUse || (WSDLUse = {}));
export var WSDLVersion;
(function (WSDLVersion) {
    WSDLVersion["WSDL_1_1"] = "1.1";
    WSDLVersion["WSDL_2_0"] = "2.0";
})(WSDLVersion || (WSDLVersion = {}));
export var WSDLErrorCode;
(function (WSDLErrorCode) {
    WSDLErrorCode["INVALID_NAMESPACE"] = "INVALID_NAMESPACE";
    WSDLErrorCode["MISSING_REFERENCE"] = "MISSING_REFERENCE";
    WSDLErrorCode["INVALID_BINDING"] = "INVALID_BINDING";
    WSDLErrorCode["DUPLICATE_DEFINITION"] = "DUPLICATE_DEFINITION";
    WSDLErrorCode["INVALID_OPERATION"] = "INVALID_OPERATION";
    WSDLErrorCode["MISSING_MESSAGE"] = "MISSING_MESSAGE";
    WSDLErrorCode["INVALID_PORT_TYPE"] = "INVALID_PORT_TYPE";
    WSDLErrorCode["CIRCULAR_REFERENCE"] = "CIRCULAR_REFERENCE";
})(WSDLErrorCode || (WSDLErrorCode = {}));
